### PR TITLE
Add lr setting

### PR DESCRIPTION
**Pull request by lm-konda**
add initial learning rate setting
and
add learning rate tuninigs (We may need to change the configuration name)

### CODE REVIEW

File: blueoil/templates/lmnet/classification.tpl.py [hunk 1]
```diff
@@ -78,7 +78,34 @@
...
+    }
+elif '{{lr_setting}}' == 'tune2':
+    NETWORK.LEARNING_RATE_KWARGS = {
+        "values": [{{learning_rate}}, {{learning_rate}} / 10, {{learning_rate}} / 100, {{learning_rate}} / 1000],
+        "boundaries": [int((step_per_epoch * (MAX_EPOCHS - 1)) * 1 / 3), int((step_per_epoch * (MAX_EPOCHS - 1)) * 2 / 3), int(step_per_epoch * (MAX_EPOCHS - 1))],
+    }
+elif '{{lr_setting}}' == 'tune3':
+    NETWORK.LEARNING_RATE_KWARGS = {
+        "values": [{{learning_rate}} / 1000, {{learning_rate}}, {{learning_rate}} / 10, {{learning_rate}} / 100, {{learning_rate}} / 1000],
+        "boundaries": [int((step_per_epoch * (MAX_EPOCHS - 1)) * 1 / 4), int((step_per_epoch * (MAX_EPOCHS - 1)) * 2 / 4), int((step_per_epoch * (MAX_EPOCHS - 1)) * 3 / 4), int(step_per_epoch * (MAX_EPOCHS - 1))],
```
Comment (ruimashita) [inline blueoil/templates/lmnet/classification.tpl.py]: The `1/4` epoch length of warmup stage looks long, I think `1` epoch is full for warmup. 
See also `4.2. CIFAR-10 and Analysis` in https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/He_Deep_Residual_Learning_CVPR_2016_paper.pdf

File: blueoil/blueoil_init.py [hunk 2]
```diff
@@ -272,6 +272,23 @@ def ask_questions():
...
+        'name': 'value',
+        'message': 'initial learning rate:',
+        'default': '0.001'
+    }
+    initial_learning_value = prompt(initial_lr_value_question)
+
+    training_lr_question = {
+        'type': 'rawlist',
+        'name': 'value',
+        'message': 'choose learning rate setting (tune1 / tune2 / tune3 / fixed):',
```
[the remark this answers is not in the document]
Comment (lm-konda) [reply blueoil/blueoil_init.py]: agree and add option description.

File: blueoil/generate_lmnet_config.py [hunk 3]
```diff
@@ -109,7 +109,7 @@ def _blueoil_to_lmnet(blueoil_config):
         "save_steps": 1000,
         "test_steps": 1000,
         "summarise_steps": 100,
-        "learning_rate": 1e-3,
+        "learning_rate": blueoil_config["trainer"]["initial_learning_rate"],
```
[the remark this answers is not in the document]
Comment (lm-konda) [reply blueoil/generate_lmnet_config.py]: I agree and moved.
But I think "under the 138 line" means training settings.
Do you have some good idea?

File: blueoil/templates/lmnet/classification.tpl.py [hunk 4]
```diff
@@ -78,7 +78,34 @@
 
 NETWORK = EasyDict()
 NETWORK.OPTIMIZER_CLASS = tf.train.MomentumOptimizer
-NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9, "learning_rate": {{learning_rate}}}
+
+NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9}
```
[the remark this answers is not in the document]
Comment (lm-konda) [reply blueoil/templates/lmnet/classification.tpl.py]: I agree and modified.

File: blueoil/templates/lmnet/classification.tpl.py (hunk 1 above)
Comment (lm-konda) [reply blueoil/templates/lmnet/classification.tpl.py]: I agree and current implementation is like below.
1 epoch , 1/3 , 2/3, (max-1) epoch

File: blueoil/templates/lmnet/classification.tpl.py [hunk 5]
```diff
@@ -78,7 +78,34 @@
...
+step_per_epoch = int(50000 / BATCH_SIZE)
```
[the remark this answers is not in the document]
Comment (lm-konda) [reply blueoil/templates/lmnet/classification.tpl.py]: I agree and changed to this implementation.
Thanks!

File: blueoil/blueoil_init.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ruimashita) [reply blueoil/blueoil_init.py]: Thanks 👍

File: blueoil/templates/lmnet/classification.tpl.py (hunk 1 above)
Comment (ruimashita) [reply blueoil/templates/lmnet/classification.tpl.py]: 👍

File: blueoil/templates/lmnet/classification.tpl.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ruimashita) [reply blueoil/templates/lmnet/classification.tpl.py]: @lm-konda 
Thanks,  but sorry, I was misleading 😭 

* When `lr_setting` is `fixed`, the parameter of `NETWORK.OPTIMIZER_KWARGS` is overridden here.
https://github.com/blue-oil/blueoil/blob/b7c93dc0c8a9075ddb03ac04bc12f28eeb94f6c1/blueoil/templates/lmnet/classification.tpl.py#L103-L104
* Only when `lr_setting` is not `fixed`, We use `NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9}`.

In order to avoid override `NETWORK.OPTIMIZER_KWARGS`,  remove the `NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9}` line 82 and add like this
```
if '{{learning_rate_setting}}' != 'fixed':
    NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9}
    NETWORK.LEARNING_RATE_FUNC = tf.train.piecewise_constant
```

It is easy for me to read the source code, How about this?

[the remark this answers is not in the document]
Comment (lm-konda) [reply blueoil/templates/lmnet/classification.tpl.py]: Thank you! I understand and simplify this.

File: blueoil/templates/lmnet/classification.tpl.py [hunk 6]
```diff
@@ -78,7 +78,33 @@
 
 NETWORK = EasyDict()
 NETWORK.OPTIMIZER_CLASS = tf.train.MomentumOptimizer
-NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9, "learning_rate": {{learning_rate}}}
+
+NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9}
+dataset_obj = DATASET_CLASS(subset="train", batch_size=1)
```
[the remark this answers is not in the document]
Comment (lm-konda) [reply blueoil/templates/lmnet/classification.tpl.py]: I see. I changed to reuse dataset_obj.
Thanks!

File: blueoil/generate_lmnet_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ruimashita) [reply blueoil/generate_lmnet_config.py]: We discussed about this, It is solved.

**Comment by ruimashita** [comment]
@lm-konda 
LGTM, thanks.
I will run the Jenkins tests.

**Comment by ruimashita** [comment]
Jenkins, ok to test

File: blueoil/blueoil_init.py [hunk 7]
```diff
@@ -272,6 +272,23 @@ def ask_questions():
...
+        'message': 'initial learning rate:',
+        'default': '0.001'
+    }
+    initial_learning_rate_value = prompt(initial_learning_rate_value_question)
+
+    training_learning_rate_question = {
+        'type': 'rawlist',
+        'name': 'value',
+        'message': 'choose learning rate setting(tune1 / tune2 / tune3 / fixed):',
+        'choices': ['tune1 -> "2 times decay"', 'tune2 -> "3 times decay"', 'tune3 -> "warm-up and 3 times decay"', 'fixed'],
```
Comment (ruimashita) [inline blueoil/blueoil_init.py]: In the CLI, the description looks good. 
screen shot 2018-11-14 at 21 34 50

But in the generated yml, `tune3 -> "warm-up and 3 times decay"` looks long, it is difficult to change.
```
$ cat config/xxxx.yml
...
...
trainer:
  batch_size: 10
  epochs: 100
  training_learning_rate: tune3 -> "warm-up and 3 times decay"
  initial_learning_rate: 0.1

....
....
```

We believe that easy change of generated yml by blueoil user will increase usability.
From that point of view, In the generated yml, User only seems to need key like `tune3`, don't need description like `tune3 -> "warm-up and 3 times decay"`.

The below pseudo code will solve this. 
```
choices_key_map = {
  'tune1 -> "2 times decay"'': 'tune1',
  'tune2-> "3 times decay"'': 'tune2',
  'tune3 -> "warm-up and 3 times decay"': 'tune3',
  'fixed': 'fixed',
}
tmp_learning_rate_setting = prompt(training_learning_rate_question)
training_learning_rate_setting = choices_key_map[tmp_learning_rate_setting]
```

File: blueoil/templates/blueoil-config.tpl.yml [hunk 8]
```diff
@@ -11,6 +11,8 @@ dataset:
 trainer:
   batch_size: {{ batch_size }}
   epochs: {{ training_epochs }}
+  training_learning_rate: {{ training_learning_rate_setting }}
```
Comment (ruimashita) [inline blueoil/templates/blueoil-config.tpl.yml]: I think `learning_rate_setting`  is better name than `training_learning_rate` .

Comment (ruimashita) [reply blueoil/templates/blueoil-config.tpl.yml]: Also, plz write comment to easy change for blueoil user.
like this: 
https://github.com/blue-oil/blueoil/blob/2a6604d7a17516aa25df8de527c516c571a95fc9/blueoil/templates/blueoil-config.tpl.yml#L1-L2

**Comment by ruimashita** [comment]
Please solve the `[blueoil] jenkins test` error of `blueoil_test.sh`.
When you run `blueoil_test.sh` in your environment the same error will come.

To modify `blueoil_test.sh`, You can reference the other PR
https://github.com/blue-oil/blueoil/pull/13/files#diff-e1165ef1f13d27e8f1f8f0fb4498aa7cR129
https://github.com/blue-oil/blueoil/pull/13/files#diff-b44335ae7249f9a84c94ea154cb44489R16

**Comment by lm-konda** [comment]
I am trying to fix blueoil_test.sh.
But I think there are some problems.

It seems that there is inconsistency in the whitespace of the automatically generated configuration file.
Is it a problem only in my environment?

I cope with ad hoc processing, but it is not elegant.

**Comment by ruimashita** [comment]
@lm-konda 

> It seems that there is inconsistency in the whitespace of the automatically generated configuration file.
Is it a problem only in my environment?

Do you talk about white-space after `0.001` of `tests/config/***.yml` ?

screen shot 2018-11-15 at 15 48 26

Maybe, `{{ initial_learning_rate_value }}  ` of `blueoil/templates/blueoil-config.tpl.yml` makes wrong.  
https://github.com/blue-oil/blueoil/blob/cec9aaa31b1a272bdbe8062e5703faf16e272f85/blueoil/templates/blueoil-config.tpl.yml#L16

File: blueoil/templates/blueoil-config.tpl.yml (hunk 8 above)
Comment (ruimashita) [reply blueoil/templates/blueoil-config.tpl.yml]: Sorry, I was mislead.

As already discussed,
The name of `learning_rate_setting` is good. 
The comment of `learning_rate_setting` is like below.

from
```
training_learning_rate: {{ training_learning_rate_setting }}
```

to
```
# supported 'learning_rate_setting' is 'tune1', 'tune2', 'tune3', 'fixed'.
#   'tune1' is 2 times decay, learning rate reduce to 1/10 on epoch/2 and epoch-1. 
#   'tune2' is 3 times decay, learning rate reduce to 1/10 on epoch/3 and epoch*2/3 and epoch-1. 
#   'tune3' is warmup and 3 times decay, warmup learning rate 1/1000 in 1 epoch, then train same as 'tune2'. 
#   'fixed' is constant learning rate.
learning_rate_setting: {{ training_learning_rate_setting }}
```

File: blueoil_test.sh [hunk 9]
```diff
@@ -211,7 +215,9 @@ if [ "${YML_CONFIG_FILE}" == "" ]; then
             do
                 TRAINING_DATASET_PATH=$(get_yaml_param train_path tests/config/${TEST_CASE}.yml)
                 VALIDATION_DATASET_PATH=$(get_yaml_param test_path tests/config/${TEST_CASE}.yml)
-                init_test ${TEST_CASE} ${TASK_TYPE_NUMBER} 1 1 ${DATASET_FORMAT_NUMBER} ${TRAINING_DATASET_PATH} ${VALIDATION_DATASET_PATH}
+                INITIAL_LEARNING_RATE=$(get_yaml_param initial_learning_rate tests/config/${TEST_CASE}.yml)
```
Comment (ruimashita) [inline blueoil_test.sh]: Do we need these changes ?

```
-                init_test ${TEST_CASE} ${TASK_TYPE_NUMBER} 1 1 ${DATASET_FORMAT_NUMBER} ${TRAINING_DATASET_PATH} ${VALIDATION_DATASET_PATH} 
+                INITIAL_LEARNING_RATE=$(get_yaml_param initial_learning_rate tests/config/${TEST_CASE}.yml)
+                TRAINING_LEARNING_RATE=$(get_yaml_param training_learning_rate tests/config/${TEST_CASE}.yml)
+		init_test ${TEST_CASE} ${TASK_TYPE_NUMBER} 1 1 ${DATASET_FORMAT_NUMBER} ${TRAINING_DATASET_PATH} ${VALIDATION_DATASET_PATH} 
```

File: blueoil/templates/blueoil-config.tpl.yml (hunk 8 above)
Comment (lm-konda) [reply blueoil/templates/blueoil-config.tpl.yml]: Thank you. I rewrite like this.

File: blueoil_test.sh (hunk 9 above)
Comment (lm-konda) [reply blueoil_test.sh]: Oh, sorry. I forgot to delete these lines.

Comment (ruimashita) [reply blueoil_test.sh]: Thanks 👍

File: blueoil/templates/blueoil-config.tpl.yml (hunk 8 above)
Comment (ruimashita) [reply blueoil/templates/blueoil-config.tpl.yml]: Thanks 👍

File: blueoil/templates/lmnet/classification.tpl.py [hunk 10]
```diff
@@ -50,7 +50,9 @@
 DATA_FORMAT = "NHWC"
 TASK = Tasks.CLASSIFICATION
 # In order to get instance property `classes`, instantiate DATASET_CLASS.
-CLASSES = DATASET_CLASS(subset="train", batch_size=1).classes
+dataset_obj = DATASET_CLASS(subset="train", batch_size=1)
+CLASSES = dataset_obj.classes
+step_per_epoch = int(dataset_obj.num_per_epoch/BATCH_SIZE)
```
[the remark this answers is not in the document]
Comment (ruimashita) [reply blueoil/templates/lmnet/classification.tpl.py]: @lm-konda 
In the case
* dataset_obj.num_per_epoch is 15
* BATCH_SIZE is 10
* MAX_EPOCHS is 100
* lr settibng: tune2

I got this results `'boundaries': [33, 66, 99]`.
But It should be `[49 (=15/10*99*1/3), 99 (=15/10*99*2/3), 148 (=15/10*99)]`

Maybe I think to remove `int()`
```
step_per_epoch = float(dataset_obj.num_per_epoch)/BATCH_SIZE
```
instead of current 
https://github.com/blue-oil/blueoil/blob/0283ddd1324d0996ce8cc09ee85c435f5a196e2d/blueoil/templates/lmnet/classification.tpl.py#L55
is correct.
How about this ?

[the remark this answers is not in the document]
Comment (lm-konda) [reply blueoil/templates/lmnet/classification.tpl.py]: Thanks! You are right. I removed int().

File: blueoil_test.sh [hunk 11]
```diff
@@ -213,7 +217,7 @@ if [ "${YML_CONFIG_FILE}" == "" ]; then
             do
                 TRAINING_DATASET_PATH=$(get_yaml_param train_path tests/config/${TEST_CASE}.yml)
                 VALIDATION_DATASET_PATH=$(get_yaml_param test_path tests/config/${TEST_CASE}.yml)
-                init_test ${TEST_CASE} ${TASK_TYPE_NUMBER} 1 1 ${DATASET_FORMAT_NUMBER} ${TRAINING_DATASET_PATH} ${VALIDATION_DATASET_PATH}
```
Comment (ruimashita) [inline blueoil_test.sh]: @lm-konda 
trivial things.
Maybe it is not need changes, But some white-space and tab is added.
Would you clean up to be no changes ?
